language/move-lang/src/unit_test/plan_builder.rs

```diff
@@ -0,0 +1,297 @@
+        ])
+    }
+
+    if test_attributes.is_empty() {
+        return None;
+    }
+
+    // Check for duplicate #[test(...)] attributes
+    if test_attributes.len() > 1 {
+        let len = test_attributes.len();
+        context.env.add_error(vec![
+            (
+                test_attributes[len - 1].loc,
+                "Duplicate '#[test(...)]' attribute. Only one #[test(...)] attribute is allowed",
+            ),
+            (test_attributes[len - 2].loc, "Previously annotated here"),
+            (fn_loc, "In this test"),
+        ])
+    }
+
+    // A #[test] function cannot also be annotated #[test_only]
+    if !test_only_attributes.is_empty() {
+        let msg = "Function annotated as both #[test(...)] and #[test_only]. You need to declare it as either one or the other";
+        context.env.add_error(vec![
+            (test_only_attributes.last().unwrap().loc, msg),
+            (
+                test_attributes.last().unwrap().loc,
+                "Previously annotated here",
+            ),
+            (fn_loc, "In this test"),
+        ])
+    }
+
+    // Only one abort can be specified
+    if abort_attributes.len() > 1 {
+        let len = abort_attributes.len();
+        context.env.add_error(vec![
+            (abort_attributes[len - 1].loc, "Duplicate '#[expected_failure]' attribute. Only one #[expected_failure] attribute is allowed"),
+            (abort_attributes[len - 2].loc, "Previously annotated here")
+        ])
+    }
+
+    let test_attribute = test_attributes.last().unwrap();
+    let test_annotation_params = parse_test_attribute(context, test_attribute);
+    let mut arguments = Vec::new();
+
+    for (var, _) in &function.signature.parameters {
+        match test_annotation_params.get(var.value()) {
+            Some(value) => arguments.push(value.clone()),
+            None => context.env.add_error(vec![
+                (
+                    test_attribute.loc,
+                    "Missing test parameter assignment in test. Expected a parameter to be assigned in this attribute",
+                ),
+                (var.loc(), "Corresponding to this variable"),
+                (fn_loc, "In this test"),
+            ]),
+        }
+    }
+
+    let expected_failure = if abort_attributes.is_empty() {
+        None
+    } else {
+        parse_failure_attribute(context, abort_attributes.last().unwrap())
+    };
+
+    Some(TestCase {
+        test_name: fn_name.to_string(),
+        arguments,
+        expected_failure,
+    })
+}
+
+//***************************************************************************
+// Attribute parsers
+//***************************************************************************
+
+fn parse_test_attribute(
+    context: &mut Context,
+    sp!(aloc, test_attribute): &E::Attribute,
+) -> BTreeMap<String, MoveValue> {
+    use E::Attribute_ as EA;
+    let mut args = BTreeMap::new();
```

Comment:
Not a big fan of having the map out here and then populated inside, as you lose a bit of intentionality when reading each branch of the match